### DISCUSSION TITLE
Upgrade:  new session are routed to the old spark driver

### DISCUSSION THREAD

**Discussion by user-GitHub-user** [apache/kyuubi#4756] [Q&A]
kyuubi version: 1.7.0
1.start node1 kyuubi server
2.connect use zk mode
3. run  ./bin/kyuubi-ctl delete server --host "node1" --port "10009"
4.quit connect
5. start node2 kyuubi server and version still 1.7.0
6. other connect use zk mode

The connection from Step 6 will use the spark driver from Step 2 ,  It looks like the connection from Step 2 didn't close automatically when I quit, and shouldn't the connection from Step 6 be using new spark driver?

Do I have to change the version？But what if I just want to migrate from yarn to k8s or local?

**Comment by pan3793** [comment]
Please read the [article](https://kyuubi.readthedocs.io/en/master/deployment/engine_share_level.html) to learn the concept of "Engine Share Level"

**Reply by user-GitHub-user** [reply]
> Please read the [article](https://kyuubi.readthedocs.io/en/master/deployment/engine_share_level.html) to learn the concept of "Engine Share Level"

Hello, I have read the content related to sharing level, but I do not understand if there is no problem in sharing spark driver under different versions. Or I changed spark mode

**Reply by pan3793** [reply]
> different versions

what do you mean about "version"? who's version? what's your expected result?

**Reply by user-GitHub-user** [reply]
> > different versions
> 
> what do you mean about "version"? who's version? what's your expected result?

is kyuubi version, eg : if i delete kyuubi server of 1.6.0 and start new kyuubi server of 1.7.0, new connect will reuse 1.6.0 spark driver, Is this normal？

**Reply by pan3793** [reply]
No, Kyuubi's version is a variable in routing rules

**Reply by user-GitHub-user** [reply]
> No, Kyuubi's version is a variable in routing rules

but if i not change the kyuubi version, new kyuubi server still use 1.6.0 ,then new connect reuse the spark driver, Even if I change spark master from yarn to local[*] in new kyuubi server , it still run in yarn

**Reply by pan3793** [reply]
it's by design, use the CONNECTION share level if you don't want the engine shared across sessions

**Reply by user-GitHub-user** [reply]
> it's by design, use the CONNECTION share level if you don't want the engine shared across sessions

I was just testing whether the upgraded function is normal, and I found this problem. Maybe it is not a big problem, but I think the priority of delete server operation should be higher than the Engine Share Level. Thank you for your answer

**Reply by pan3793** [reply]
To be clear, the engine's lifecycle is not managed by the server, the engine will self-terminate when there is no active session and exceeds the idle time threshold.

**Comment by pan3793** [comment]
> what if I just want to migrate from yarn to k8s or local

Kyuubi respects your Spark Client and Spark Configurations, please read the [doc](https://kyuubi.readthedocs.io/en/master/deployment/settings.html#spark-configurations) to learn how to manage Spark configurations. And you can check the whole `spark-submit` command in Kyuubi Server's log
